include/tables/FileProvenanceXAttrBufferTable.h

```diff
@@ -177,69 +311,67 @@ class FileProvenanceXAttrBufferTable : public DBTable<FPXAttrBufferRowPart> {
-    a[1] = key.mNamespace;
-    a[2] = key.mName;
-    FPXAttrBufferRowPartVec partsVec = doRead(connection, "xattr_versions", a, key.mInodeId);
-    return combine(partsVec);
-  }
-
-  private:
-  inline static bool readCheckExists(FPXAttrBufferPK key, FPXAttrBufferRow row) {
-    return key.mInodeId == row.mInodeId && key.mNamespace == row.mNamespace 
-    && key.mName == row.mName && key.mInodeLogicalTime == row.mInodeLogicalTime;
-  }
-
-  FPXAttrBufferRow get(Ndb* connection, Int64 inodeId, Int8 ns, std::string name, int inodeLogicalTime, int numParts) {
-    LOG_DEBUG("FileProvXAttr get by parts " << numParts);
-    AnyVec anyVec;
-    for(Int16 index=0; index < numParts; index++){
-       AnyMap a;
-       a[0] = inodeId;
-       a[1] = ns;
-       a[2] = name;
-       a[3] = inodeLogicalTime;
-       a[4] = index;
-       anyVec.push_back(a);
-    }
-    FPXAttrBufferRowPartVec parts = doRead(connection,anyVec);
-    LOG_DEBUG("FileProvXAttr batch read parts " << parts.size());
-    return FPXAttrBufferRow(parts);
-  }
+  /**
+   * Get a range of prov cores between fromLogicalTime and toKey.mInodeLogicalTime.
+   * This range is sparse (doesn't contain an entry for each key), but we also don't know the exact keys that have a value
+   * The values themselves are also made of multiple parts and we only know the size of the parts after reading part0 of it
+   * @param connection
+   * @param toKey
+   * @param fromLogicalTime
+   * @return
+   */
+  std::map<int, boost::optional<FPXAttrBufferRow>> getProvCore(Ndb* connection, Int64 inodeId, int fromLogicalTime, int toLogicalTime) {
+    std::map<int, boost::optional<FPXAttrBufferRow>> result;
 
-  typedef boost::unordered_map<std::string, FPXAttrBufferRowPartVec> FPXAttrBufferRowPartMap;
+    //generate part0 keys for interval fromLogicalTime to toKey.mInodeLogicalTime
+    std::vector<FPXAttrBufferPK> keys;
+    AnyVec keysVec;
+    for(int logicalTime=fromLogicalTime; logicalTime <= toLogicalTime; logicalTime++){
+      FPXAttrBufferPK key(inodeId, FileProvenanceConstantsRaw::XATTRS_USER_NAMESPACE, FileProvenanceConstantsRaw::XATTR_PROV_CORE, logicalTime, 0);
+      keys.push_back(key);
+      keysVec.push_back(key.getKey0Map());
+    }
+    std::vector<FPXAttrBufferRowPart> rows = DBTable<FPXAttrBufferRowPart>::doRead(connection, keysVec);
 
-  std::vector<FPXAttrBufferRow> combine(FPXAttrBufferRowPartVec& partsVec){ 
-    LOG_DEBUG("FileProvXAttr combine parts " << partsVec.size());   
-    FPXAttrBufferRowPartMap xattrsMap;
-    for(auto& part: partsVec){
-      std::string id = part.getUniqueXAttrId();
-      if(xattrsMap.find(id) == xattrsMap.end()){
-        xattrsMap[id] = FPXAttrBufferRowPartVec();
+    //save prov cores with only 1 part directly to result and generate the keys for multi part values
+    AnyVec multiPartKeyVec;
+    //holds for each prov core key0 and numParts
+    std::vector<FPXAttrBufferPK> multiPartKeys;
+    for(unsigned i=0; i<keys.size(); i++){
```

Comment:
i think we should use unsigned int to be consistent

Reply:
fixed